MFEM: add new version v4.7

v-dobrev opened the pull request:
TODO:
- [x] After MFEM v4.7 is released, update the source for that in `mfem/package.py`
- [x] Add new variant for `precision`: `double`, or `single`; available since v4.7
- [x] Address #41673
- [x] Address https://github.com/spack/spack/pull/44427#issuecomment-2137970683

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

v-dobrev committed:
[mfem] Initial changes for mfem-4.7

v-dobrev committed:
[mfem] remove extra space from branch name

v-dobrev committed:
[suite-sparse] edits:
* Fix the build when the cuda variant is disabled and cuda is
  present on the system.
* In the 'libs' property, search for the libraries in the whole
  prefix, not just in 'lib' -- newer versions use 'lib64' at least
  in some cases.
* Propagate the 'make_jobs' setting to the build system -- it
  requires a special flag in the make options.

v-dobrev committed:
[mfem] New versions of 'umpire' may require linking with 'fmt'

v-dobrev committed:
[mfem] Updates in the testing script for v4.7

v-dobrev committed:
[mfem] Replace mfem-4.7-dev branch with the official release link

v-dobrev committed:
[hiop] Some correction in the HiOp package

v-dobrev committed:
[build_environment.py] In get_rpaths(pkg), do not add the result of
path_from_modules(...) to the rpaths when it returns None. The presense
of None in the rpaths was causing an issue.

v-dobrev committed:
[metis] When compiling 'metis@4.0.3' with 'cce@17:', add the flag
'-std=c89' -- without it the build fails.

v-dobrev committed:
[zlib-ng] Fix an issue in the 'libs' property, descibed in a comment in
the code.

v-dobrev committed:
[petsc] Fix an issue where deprecation warnings from HIP/ROCm headers was
causing the PETSc configuration to fail.

v-dobrev committed:
[suite-sparse] Propagate the settings of the variants 'cuda' and 'openmp'
to the SuiteSparse build system. Add backward compatibility support for
new SuiteSparse versions which moved the SuiteSparse headers inside a
subdirectory 'suitesparse' -- to allow existing packages to work with these
new versions, all headers from the 'suitesparse' subdirectory are linked in
the parent directory, '<prefix>/include'. Added new version: v7.7.0.

v-dobrev committed:
[mfem] Added a patch for mfem-4.7 to fix a few small issues noticed
during testing through Spack. Added various tweaks and fixes in the
package and updated the testing script 'test_builds.sh'.

v-dobrev committed:
Merge branch 'develop' into packages/mfem-4.7

v-dobrev commented:
@spackbot fix style

v-dobrev committed:
[@spackbot] updating style on behalf of v-dobrev

v-dobrev committed:
[hypre] Add 'precision' variant

v-dobrev committed:
[mfem] Add 'precision' variant

v-dobrev committed:
[mfem] Add 'cxxstd' variant and other updates to address #41673

v-dobrev committed:
[@spackbot] updating style on behalf of v-dobrev

v-dobrev committed:
[mfem] define the functions 'ld_flags_from_library_list' and
'ld_flags_from_dirs' as methods of class 'Mfem' so that they
can be reused by sub-classes.

v-dobrev commented:
@spackbot fix style

v-dobrev committed:
[@spackbot] updating style on behalf of v-dobrev

kennyweiss reviewed (dismissed):
Thanks @v-dobrev!

scheibelp commented on var/spack/repos/builtin/packages/zlib-ng/package.py:
```diff
@@ -69,9 +69,13 @@ class ZlibNg(AutotoolsPackage, CMakePackage):
+        shared = self.spec.satisfies("+shared")
+        libs = find_libraries(name, root=self.prefix, recursive=True, shared=shared)
+        # With some compilers, e.g. cce@17, building a shared library may fail,
+        # so the build system just builds a static version.
```
Ideally we'd fail at install time for `zlib-ng` if it didn't have the libs we expect (as it is, this is working around a case where the spec does not actually represent what is installed), and we could add directives like `conflicts(+shared, when=%cce@17)` to account for known cases where we need to build static libs.

Do you think you could change that here?

v-dobrev replied:
The issue here is that `cce@17` uses the LLD linker which fails the autodetection of `zlib-ng` of how to build a shared library. A better workaround than what I have now, is to have the `zlib-ng` package modify the `ldflags` by adding `-Wl,--undefined-version` when using `cce@17` -- older versions of `cce` may be affected too but I'm not sure when it started using LLD.

If you prefer this, I can replace the current workaround with the above -- what's the easiest way to modify the `ldflags` in `zlib-ng`?

scheibelp replied:
> what's the easiest way to modify the ldflags in zlib-ng?

To answer this directly: there is an existing `flag_handler` method - you can add a case for `ldflags`/`cce@17`:

```
    def flag_handler(self, name, flags):
        ...
        if name == "ldflags" and self.spec.satisfies("%cce@17"):
            flags.append("-Wl,--undefined-version")
        return (flags, None, None)
```

> uses the LLD linker which fails the autodetection of zlib-ng of how to build a shared library.

I'd be hopeful there might be some way to help it succeed at this. Then we could potentially address other cases of compilers using `lld`. That doesn't have to happen here, but the error message that comes up might be useful for future reference.

v-dobrev replied:
The error reported by `zlib-ng`'s confugure script is:
```console
Checking for shared library support... No shared library support; try without defining CC and CFLAGS
```
The error message from the failed command in `configure.log` is:
```console
ld.lld: error: version script assignment of 'ZLIB_1.2.0' to symbol 'compressBound' failed: symbol not defined
```
The latter error message is repeated a number of times with different versions and symbols.

Adding the flag `-Wl,--undefined-version` is basically telling the linker to ignore this specific type of errors.

Another package that needed this flag is `util-linux-uuid` -- without it the build just failed with similar error message. At the moment, I'm just setting the `ldflags` for the compiler to include this flag for all packages. To get some of the `strumpack` executables to link (but probably not actually work), I also added `-Wl,--allow-shlib-undefined` to the compiler `ldflags`. For me this combination of `ldflags: -Wl,--undefined-version,--allow-shlib-undefined` was sufficient to resolve the LLD issues I encountered.

v-dobrev replied:
Addressed in 1dbc0f61f6.

scheibelp replied:
> Another package that needed this flag is util-linux-uuid -- without it the build just failed with similar error message. 

I don't see that in https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/util-linux-uuid/package.py, or a recent PR on it. Was that in the works? Or did I miss something?

I think it might be worth following up with looking into how to avoid this, because I wonder if a `zlib-ng` library built without version definitions would potentially break other packages that would expect otherwise. Not something that has to be addressed here though.

v-dobrev replied:
I haven't proposed a fix for `util-linux-uuid` for building with LLD or specifically with `cce` -- I just noticed in my test of this PR and resolved the issue locally by setting `ldflags: -Wl,--undefined-version,--allow-shlib-undefined` in the `cce@17.0.1` compiler definition I have in `etc/spack/compilers.yaml`.

If you want, I can add the fix for that from `zlib-ng` to `util-linux-uuid` as well.

scheibelp commented on var/spack/repos/builtin/packages/suite-sparse/package.py:
```diff
@@ -236,13 +241,25 @@ def install(self, spec, prefix):
+                + " -DSUITESPARSE_USE_OPENMP=%s" % ("ON" if "+openmp" in spec else "OFF")
+                # Older versions use the negative flag NOPENMP: "OFF" means use
+                # OpenMP:
+                + " -DNOPENMP=%s" % ("OFF" if "+openmp" in spec else "ON")
```
(minor) It would be clearer IMO to break this off into an if statement based on version:

```
if self.satisfies("@:upper_bound"):
  # old option name
else
  # new option name
```

v-dobrev replied:
Addressed in b5920de.

scheibelp commented on var/spack/repos/builtin/packages/suite-sparse/package.py:
```diff
@@ -236,13 +241,25 @@ def install(self, spec, prefix):
+            # Assume the proper Cray CCE module (cce) is loaded:
+            craylibs_path = env["CRAYLIBS_" + machine().upper()]
+            env.setdefault("LDFLAGS", "")
+            env["LDFLAGS"] += " -Wl,-rpath," + craylibs_path
```
Ideally:

* this would guard against the possibility that the env variable is not set
* this would use `str(spec.target)` vs. `platform.machine()`

Also, this seems like a directory that should be RPATHed for any artifact built by `%cce` (not just `suite-sparse` in particular) - do you agree? If so, we have logic that's supposed to gather up such paths, and it's worth investigating why that didn't occur for you.

v-dobrev replied:
Unfortunately, `spec.target` is not good -- e.g. it gives `zen3` and `machine()` is `x86_64` -- I need something that returns `x86_64` or `aarch64` since the `cce` module defines `CRAYLIBS_X86_64` and `CRAYLIBS_AARCH64`.

If the environment variable is not defined, we get an exception, right? I though this is good enough. What guard do you have in mind?

I've added this logic (in previous PRs) in a few packages where I have needed this tweak before -- `strumpack`, `butterflypack`, and in `mfem` itself. Without explicitly adding this `rpath`, what happens is that the `cce` compiler links the libraries of `suite-sparse` with some libraries in the path from `CRAYLIBS_*` but does not add that path to the `rpath` (when linking a library). When linking an executable `cce` adds automatically the path from `CRAYLIBS_*` to the `rpath` and links again with the same libraries in that path, so there is no issue when the final linking is done by `cce`. However, when building `+rocm`, `mfem` compiles and links everything with `hipcc`, so the path from `CRAYLIBS_*` needs to be added explicitly as `rpath` and linking should also link with the needed `cce` libraries. The `mfem` package does that here: https://github.com/spack/spack/blob/b5920de26a55a64531cccd6c6f146902b5e599a4/var/spack/repos/builtin/packages/mfem/package.py#L987-L1001

Unfortunately, `hipcc` adds the `rpath` as `RUNPATH` and not as `RPATH` and thus the `rpath` from `CRAYLIBS_*` is not considered when loading libraries needed by mfem executables (examples or miniapps) like the `suite-sparse` libraries. This leads to "library not found" errors at runtime.

Maybe one option that I can try in the `mfem` package is to add the `CRAYLIBS_*` path while explicitly disabling the new dtags with `--disable-new-dtags` -- I think this will tell `hipcc` to add the `rpath` arguments as `RPATH` which should propagate to dependencies like `suite-sparse`. If this works, I can probably remove the logic about `CRAYLIBS_*` from other packages -- I'm not sure if that will break anything outside of `mfem` though, e.g. the `xsdk-examples`.

scheibelp replied:
> Unfortunately, spec.target is not good -- e.g. it gives zen3 and machine() is x86_64 -- I need something that returns x86_64 or aarch64

`str(spec.target.family)` will do that for you

> If the environment variable is not defined, we get an exception, right? I though this is good enough. What guard do you have in mind?

I was thinking it might be good to avoid a key error if `CRAYLIBS_*` isn't in `env` so like:

```
craylibs_path = env.get("CRAYLIBS_" + machine().upper(), None)
# Maybe this should fail if craylibs_path is None, but we can throw a clearer exception than keyerror
# like "this variable needs to be defined and isn't"
```

> When linking an executable cce adds automatically the path from CRAYLIBS_* to the rpath and links again with the same libraries in that path, so there is no issue when the final linking is done by cce. However, when building +rocm, mfem compiles and links everything with hipcc, so the path from CRAYLIBS_* needs to be added explicitly as rpath 

Thanks for that explanation: I think it makes sense to me why we wouldn't have auto-added an rpath argument

> Maybe one option that I can try in the mfem package is to add the CRAYLIBS_* path while explicitly disabling the new dtags with --disable-new-dtags -- I think this will tell hipcc to add the rpath arguments as RPATH which should propagate to dependencies like suite-sparse. 

I'd be curious if that works (that would help me think on how to improve this in general), but I think they way you have it now is preferable.

I'm not clear how this would propagate to dependencies: generally changes applied to a parent won't propagate (although there are ways to do that). FWIW the spack compiler wrappers generally add `--disable-new-dtags` automatically. If we could get the `hip` compiler to use the spack wrappers, this should happen "for free" (but I don't have a notion of whether that would be easy, or have unintended side effects).

v-dobrev replied:
Addressed the handling of the `CRAYLIBS_*` variable in f02270302d.

v-dobrev replied:
I tried commenting out the `if "%cce" in spec` section in `suite-sparse` and building with testing `mfem +rocm +suite-sparse` without modifying the mfem package (in particular without adding `--disable-new-dtags` in the `cce` handling). It turns out that actually works now -- I think because all the implicit Cray libraries that `cce` links with are now also added by mfem here: https://github.com/spack/spack/blob/b5920de26a55a64531cccd6c6f146902b5e599a4/var/spack/repos/builtin/packages/mfem/package.py#L987-L1001

I think, this did not work before because I did not have all the needed Cray libraries, e.g. I added `libpgas-shmem` later. I suspect that if these libs are explicitly linked with when building an executable with proper `RPATH` or `RUNPATH` then they are found as dependencies of the executable and if needed as dependencies-of-dependencies, they they don't need to be re-resolved.

On the other hand, however, when the `if "%cce" in spec` section in `suite-sparse` is commented out, the `suite-sparse` libraries are not self contained in the sense that `ldd libumfpack.so` shows "not found" for the Cray libraries that live in the `CRAYLIBS_*` path. For this reason, I think it's better to keep adding the extra `rpath` when building with `cce`.

scheibelp commented on lib/spack/spack/build_environment.py:
```diff
@@ -747,7 +747,9 @@ def get_rpaths(pkg):
     if pkg.compiler.modules and len(pkg.compiler.modules) > 1:
-        rpaths.append(path_from_modules([pkg.compiler.modules[1]]))
+        mod_rpath = path_from_modules([pkg.compiler.modules[1]])
+        if mod_rpath is not None:
```
(minor) might want to test `if mod_rpath` (to filter `""` as well).

v-dobrev replied:
Addressed in c8f2b95.

scheibelp commented on var/spack/repos/builtin/packages/hypre/package.py:
```diff
@@ -85,6 +85,13 @@ class Hypre(AutotoolsPackage, CudaPackage, ROCmPackage):
+        values=("single", "double", "longdouble"),
+        multi=False,
+        description="Floating point precision",
+    )
```
(question) was this always available for `hypre`, or only starting with a particular version like mfem?

v-dobrev replied:
I think it has been available in hypre for a long time, just the package did not have it as a variant.

I checked -- the precision options were added in hypre v2.12.1: `--enable-single` and `--enable-longdouble`.

I though all `hypre` versions supported by the package have these but it turns out some of the oldest ones don't. Do you think it is worth adding the constraint `when="@2.12.1:"`?

scheibelp replied:
> Do you think it is worth adding the constraint when="@2.12.1:"?

Since there are some listed `hypre` versions before this, that would be useful (although I acknowledge that it's unlikely folks will be building these old versions, we try not to get in the way of doing so).

v-dobrev replied:
Addressed in 1eaf8a663e.

scheibelp commented on var/spack/repos/builtin/packages/metis/package.py:
```diff
@@ -98,12 +98,32 @@ def setup_build_environment(self, env):
-            options.append("COPTIONS={0}".format(self.pkg.compiler.cc_pic_flag))
+            options.append(self.pkg.compiler.cc_pic_flag)
+        if self.spec.satisfies("%cce@17:"):
+            options.append("-std=c89")
```
(question) do older versions of `%cce` need a different `-std`, or are you adding this bound because you haven't used `%cce:16`?

v-dobrev replied:
I think this is only needed with the LLVM-based `cce` versions and only versions that use LLVM >= X, where X is 15 or 14, I'm not exactly sure. It's the LLVM version that made implicit function declarations in C errors, if I remember correctly.

scheibelp replied:
Thanks! It sounds like this could be expanded, but in this case IMO it's not worth doing so until someone has a chance to try with those versions.

scheibelp commented on var/spack/repos/builtin/packages/mfem/package.py:
```diff
@@ -448,6 +474,18 @@ class Mfem(Package, CudaPackage, ROCmPackage):
             "amgx~mpi cuda_arch={0}".format(sm_), when="+amgx~mpi cuda_arch={0}".format(sm_)
         )
 
+    # double precision requirements, may be incomplete:
```
(question) in what sense do you mean incomplete? `["@:4.6", "precision=double"]` appears to partition the space of possibilities based on what you added. Do you mean there might be some versions <= 4.6.... that would want to use `hypre precision=single`?

v-dobrev replied:
The "incomplete" refers to the list of dependencies to which we need to propagate the "precision=double" setting -- I did not dig deeper to find all optional packages that can be built with precision other than `double`.

scheibelp replied:
Thanks, could you add that detail to the comment? I assume the idea is that `mfem` might benefit from additional dependencies building with different precisions, but for now it doesn't seem to need this. If I misunderstood, and this isn't for `mfem`'s benefit, then IMO it would be clearer to remove the comment.

v-dobrev replied:
I moved and extended the comments (for both `double` and `single` precision) just before the respective `depends_on(...)` directives in b5a729faec. Does this make things clearer?

scheibelp replied:
(minor) Thanks but:

```
        # This set of depens_on() directives enforces some, but not all,
        # requirements in optional packages in order to have a successful
        # build. This incompleteness may cause certain specs to concretize
        # fine but fail at build time.
```

this is also confusing to me FWIW - I'm not sure what to get out of it (or the original comment)

I'd recommend something like:

```
# May need to enforce precision consistency on other packages in the future
```

v-dobrev replied:
Ok, your suggestion sounds good to me too -- I updated the comment in 0c61c4ded4.

scheibelp commented on var/spack/repos/builtin/packages/mfem/package.py:
```diff
@@ -489,56 +530,20 @@ def setup_build_environment(self, env):
-        # Also does not add system library paths as defined by 'sys_lib_paths'
-        # above -- this is done to avoid issues like this:
-        # https://github.com/mfem/mfem/issues/1088.
         def ld_flags_from_library_list(libs_list):
```
(minor) it looks like these definitions could be removed entirely (and whatever calls them could directly call `self.ld_flags_from_library_list`).

v-dobrev replied:
I specifically added these local functions, so that I didn't need to add the `self.` in front of all the calls we already had. Is there another way to avoid adding the `self.`?

scheibelp replied:
You could say `ld_flags_from_library_list = self. ld_flags_from_library_list` (assign it like a variable), which would be a bit clearer on what the intent is.

v-dobrev replied:
This is neat. Added in ef48e6ad6b.

scheibelp commented on var/spack/repos/builtin/packages/mfem/package.py:
```diff
@@ -1281,3 +1320,46 @@ def test_mk(self):
+            for dir in libs_list.directories
+            if not self.is_sys_lib_path(dir)
+        ]
+        flags += ["-L%s" % dir for dir in libs_list.directories if not self.is_sys_lib_path(dir)]
```
Could you use `spack.util.environment.filter_system_paths`?

v-dobrev replied:
We probably could. I remember that someone explicitly added the path `/usr/lib/x86_64-linux-gnu` to `mfem`'s `sys_lib_paths` (I suspect because they needed it) -- it's one path that is not included in the list used by `filter_system_paths`, so we'll probably want to filter one more time if we use `filter_system_paths`.

scheibelp replied:
> I remember that someone explicitly added the path /usr/lib/x86_64-linux-gnu to mfem's sys_lib_paths

I missed that - we should provide a convenience method for that. Until then I don't think it makes sense to ask for that refactor (including in this PR).

scheibelp reviewed (changes requested):
Most comments are minor and can be disregarded.

My main issue is with handling of libs in `zlib-ng`. I'm also concerned about handling of `CRAYLIBS_` in `suite-sparse`, primarily because I get the impression we should be handling this at a higher level, although if this fixes things for you in the meantime I'm fine with merging as is and investigating separately.

v-dobrev committed:
[spack] 'ldflags' should not be passed to CMAKE_STATIC_LINKER_FLAGS

v-dobrev committed:
[mfem] Update the MFEM testing script

v-dobrev committed:
Merge branch 'develop' into packages/mfem-4.7

Resolved conflicts:
  var/spack/repos/builtin/packages/suite-sparse/package.py

v-dobrev committed:
[suite-sparse] Address reviewer feedback

v-dobrev committed:
[spack] Address reviewer feedback:
        https://github.com/spack/spack/pull/44010#discussion_r1623752509

v-dobrev committed:
[zlib-ng] Address reviewer feedback:
          https://github.com/spack/spack/pull/44010/files#r1623078771

v-dobrev committed:
In the packages butterflaypack, mfem, strumpack, and suite-sparse,
update the code for handling the 'cce' compiler:
* use str(spec.target.family) instead of platform.machine(),
* raise an InstallError exception, with a recommendation message, if
  'CRAYLIBS_*' is not found in the environment.

v-dobrev committed:
[hypre] Add minimum version requirement for the 'precision' variant

v-dobrev committed:
[mfem] Address reviewer feedback:
       https://github.com/spack/spack/pull/44010#discussion_r1623762711

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -162,7 +162,9 @@ def initconfig_compiler_entries(self):
-            for ld_type in ["EXE", "MODULE", "SHARED", "STATIC"]:
+            # 'ldflags' should not be used with CMAKE_STATIC_LINKER_FLAGS which
+            # is used by the archiver, so don't include "STATIC" in this loop:
+            for ld_type in ["EXE", "MODULE", "SHARED"]:
```
I'm curious if @white238 agrees with this

v-dobrev replied:
For some context: I ran into this issue when building RAJA and I had `ldflags` setup to be `-Wl,--undefined-version`. Then I got a failure in the following archive command from cmake:
```console
/usr/bin/llvm-ar qc lib/libRAJA.a -Wl,--undefined-version CMakeFiles/RAJA.dir/src/AlignedRangeIndexSetBuilders.cpp.o ...
```
After some digging, I found that `CMAKE_STATIC_LINKER_FLAGS` was set to `-Wl,--undefined-version` which was causing it to appear in the above command. From there, I tracked the place where `CMAKE_STATIC_LINKER_FLAGS` was set to the above code.

white238 replied:
I would agree @v-dobrev is correct in this fix and this CMake page backs that up.

https://cmake.org/cmake/help/latest/envvar/LDFLAGS.html

Though this opens a different question, that if we want to set a global `CMAKE_STATIC_LINKER_FLAGS`, I am not sure how we would do that with Spack's compiler flags. @becker33, thoughts?

scheibelp replied:
For flags that we want to apply for `ar`

* we could introduce another `*flags` property (alongside `cflags` etc.)
* or we could add logic to filter out all flags that are not compatible with `ar` from `ldflags` (I think the first suggestion is more clear though)

scheibelp reviewed (dismissed):
I agree with all changes but don't quite understand the implications of https://github.com/spack/spack/pull/44010/files#r1628611537, so I pinged the person who wrote that bit (and I'd like to wait up to a day or two to see if they agree with it, or investigate it more myself if they are busy).

v-dobrev committed:
[mfem] Address reviewer feedback:
       https://github.com/spack/spack/pull/44010#discussion_r1623761626

v-dobrev commented:
@ulrikeyang, @osborn9,

I'm adding a `precision` variant (`double` - default, `single`, or `longdouble`) to the hypre package -- this was need to support the new `precision` variant in mfem.

Let me know if you want me to make any adjustments, or feel free to make them yourself in this PR directly.

v-dobrev commented:
@cnpetra, @cameronrutherford,

I made some changes to the HiOp package -- I needed these so that I can successfully build and test various builds of mfem with hiop.

Let me know if you want me to make any adjustments in the hiop package, or feel free to make them yourself in this PR directly.

v-dobrev commented on var/spack/repos/builtin/packages/mfem/package.py:
```diff
@@ -468,7 +511,10 @@ class Mfem(Package, CudaPackage, ROCmPackage):
         sha256="2a31682d876626529e2778a216d403648b83b90997873659a505d982d0e65beb",
     )
-    patch("mfem-hip.patch", when="+rocm ^hip@6.0:")
+    # FIXME: if we need this patch for v4.7.0, we need a new patch since this
+    #        one fails
+    patch("mfem-hip.patch", when="@4.6.0 +rocm ^hip@6.0:")
```
Hi @renjithravindrankannath, it looks like you added this patch in commit 118ff338a, so I wanted to understand the issue it fixes and figure out if we need a new patch for mfem v4.7.

It looks like with HIP >= 6, the patch and the other change in 118ff338a add explicit linking with `libamdhip64` which seems strange -- isn't that library automatically linked with when using `hipcc` (for compiling and/or linking) as does mfem?

cnpetra replied:
@nychiang can you please take a look?

renjithravindrankannath replied:
> Hi @renjithravindrankannath, it looks like you added this patch in commit [118ff33](https://github.com/spack/spack/commit/118ff338abbf242712b822f2bffa954eb46654e5), so I wanted to understand the issue it fixes and figure out if we need a new patch for mfem v4.7.
> 
> It looks like with HIP >= 6, the patch and the other change in [118ff33](https://github.com/spack/spack/commit/118ff338abbf242712b822f2bffa954eb46654e5) add explicit linking with `libamdhip64` which seems strange -- isn't that library automatically linked with when using `hipcc` (for compiling and/or linking) as does mfem?

Hi @v-dobrev ,
This patch was added when it was unable to find libhsa-runtime64.so and load libamdhip64.so. [PR 42876](https://github.com/spack/spack/pull/42876) fixed the issue and patch can be removed for 6.1.

v-dobrev replied:
> @nychiang can you please take a look?

@cnpetra, @nychiang, just FYI: this particular discussion in the PR is not related to HiOp -- you probably just want to look at the changes in the HiOp Spack package.

v-dobrev replied:
> > Hi @renjithravindrankannath, it looks like you added this patch in commit [118ff33](https://github.com/spack/spack/commit/118ff338abbf242712b822f2bffa954eb46654e5), so I wanted to understand the issue it fixes and figure out if we need a new patch for mfem v4.7.
> > It looks like with HIP >= 6, the patch and the other change in [118ff33](https://github.com/spack/spack/commit/118ff338abbf242712b822f2bffa954eb46654e5) add explicit linking with `libamdhip64` which seems strange -- isn't that library automatically linked with when using `hipcc` (for compiling and/or linking) as does mfem?
> 
> Hi @v-dobrev , This patch was added when it was unable to find libhsa-runtime64.so and load libamdhip64.so. [PR 42876](https://github.com/spack/spack/pull/42876) fixed the issue and patch can be removed for 6.1.

Thanks for the clarification, @renjithravindrankannath. If I understand correctly the Spack PR you linked (#42876) also fixes the issue for HIP v6.0 (by applying a patch present in HIP v6.1), so it seems in the latest Spack we don't need the patch you added to MFEM neither for HIP v6.0 nor for HIP v6.1. Do you agree? Or did I misunderstand?

v-dobrev replied:
Looking again at #42876, it looks like the patch there is applied only when using Spack-built HIP v6.0. If using external HIP v6.0, the issue may still be present. I'll try to test with external HIP v6.0 and add an updated patch for mfem v4.7, if needed.

renjithravindrankannath replied:
> > > Hi @renjithravindrankannath, it looks like you added this patch in commit [118ff33](https://github.com/spack/spack/commit/118ff338abbf242712b822f2bffa954eb46654e5), so I wanted to understand the issue it fixes and figure out if we need a new patch for mfem v4.7.
> > > It looks like with HIP >= 6, the patch and the other change in [118ff33](https://github.com/spack/spack/commit/118ff338abbf242712b822f2bffa954eb46654e5) add explicit linking with `libamdhip64` which seems strange -- isn't that library automatically linked with when using `hipcc` (for compiling and/or linking) as does mfem?
> > 
> > 
> > Hi @v-dobrev , This patch was added when it was unable to find libhsa-runtime64.so and load libamdhip64.so. [PR 42876](https://github.com/spack/spack/pull/42876) fixed the issue and patch can be removed for 6.1.
> 
> Thanks for the clarification, @renjithravindrankannath. If I understand correctly the Spack PR you linked (#42876) also fixes the issue for HIP v6.0 (by applying a patch present in HIP v6.1), so it seems in the latest Spack we don't need the patch you added to MFEM neither for HIP v6.0 nor for HIP v6.1. Do you agree? Or did I misunderstand?

Hi @v-dobrev , Yes, You are right. Both 6.0 and 6.1 doesn't need the patch.

v-dobrev replied:
I tested with external hip-6.0.0 and the without the `mfem-hip.patch` -- both mfem-4.6.0 and mfem-4.7.0 worked fine. Therefore, I decided to remove the patch altogether, see 05d4e606da.

cameronrutherford commented on var/spack/repos/builtin/packages/hiop/package.py:
```diff
@@ -134,9 +138,10 @@ class Hiop(CMakePackage, CudaPackage, ROCmPackage):
+    depends_on("raja@0.14.0:0.14", when="@0.3.99:+raja")
+    depends_on("umpire@6.0.0:6", when="@0.3.99:+raja")
+    depends_on("camp@0.2.3:0.2", when="@0.3.99:+raja")
+    conflicts("+raja", when="@:0.3.0", msg="RAJA support requires HiOp > v0.3")
```
I think we can just change the RAJA variant to be conditional instead of this conflict.

`variant("raja", default=False, when="@:0.3.0", description="Enable/Disable RAJA")`

v-dobrev replied:
I suspect you meant `when="@0.3.99:"`. Added in 11df808814.

cameronrutherford replied:
Yes I did 🥲 thank you

cameronrutherford commented on var/spack/repos/builtin/packages/hiop/package.py:
```diff
@@ -134,9 +138,10 @@ class Hiop(CMakePackage, CudaPackage, ROCmPackage):
-    depends_on("raja@0.14.0:0.14", when="@0.5.0:+raja")
-    depends_on("umpire@6.0.0:6", when="@0.5.0:+raja")
-    depends_on("camp@0.2.3:0.2", when="@0.5.0:+raja")
+    depends_on("raja@0.14.0:0.14", when="@0.3.99:+raja")
+    depends_on("umpire@6.0.0:6", when="@0.3.99:+raja")
+    depends_on("camp@0.2.3:0.2", when="@0.3.99:+raja")
```
```suggestion
    depends_on("raja@0.14.0:0.14", when="@0.5.0:+raja")
    depends_on("umpire@6.0.0:6", when="@0.5.0:+raja")
    depends_on("camp@0.2.3:0.2", when="@0.3.99:+raja")
    depends_on("raja@:0.13", when="@0.3.99:+raja")
    depends_on("umpire@:5", when="@0.3.99:+raja")
```

(I don't think that this syntax is correct, but I am trying to communicate what I am putting in english below)

`hiop@0.3.99` requires RAJA and uses C++11 as the C++ standard. HiOp introduced C++14 as the standard, and so after that release, we are compatible with `raja@0.14.0:0.14`. RAJA 0.14.0 introduced C++14 as a required standard for consuming projects, hence this also required a change in HiOp.

I am unsure about the camp dependency...

v-dobrev replied:
Addressed in 11df808814. Take another look to see if I understood you correctly.

cameronrutherford commented on var/spack/repos/builtin/packages/hiop/package.py:
```diff
@@ -149,8 +154,10 @@ class Hiop(CMakePackage, CudaPackage, ROCmPackage):
 
     # We rely on RAJA / Umpire utilities when supporting CUDA backend
     conflicts("~raja", when="+cuda", msg="RAJA is required for CUDA support")
+    conflicts("~raja", when="+rocm", msg="RAJA is required for ROCm support")
```
I don't know how we missed this bug... Thanks for catching it!

cameronrutherford reviewed (changes requested):
Some small changes necessary to HiOp, but LGTM

v-dobrev committed:
[hiop] Address reviewer feedback:
       https://github.com/spack/spack/pull/44010#pullrequestreview-2102551504

cameronrutherford reviewed (approved):
LGTM for HiOp changes

v-dobrev committed:
[mfem] Remove HIP patch that is no longer needed.

v-dobrev committed:
[suite-sparse] Fix the CUDA flag name for versions > 7.3.1

v-dobrev committed:
[mfem] Fix typo '-Xcompiler' -> '-Xlinker'

scheibelp reviewed (dismissed):
Thanks for the edits. I'm approving on my end. Feel free to close https://github.com/spack/spack/pull/44010#discussion_r1633592130 if you disagree. It seemed like you were also checking in with other folks, so I'll wait to merge until you mention that you think this is ready.

v-dobrev committed:
[mfem] Update comments based on reviewer feedback:
       https://github.com/spack/spack/pull/44010#discussion_r1633592130

v-dobrev committed:
[mfem] Add a fix for the Umpire unit test to mfem-4.7.patch

tzanio reviewed (approved):
Thanks @v-dobrev !

v-dobrev commented:
@scheibelp, with your approval and that of @tzanio, I think we are ready to merge this.

haampie commented on var/spack/repos/builtin/packages/zlib-ng/package.py:
```diff
@@ -76,11 +76,12 @@ def libs(self):
     def flag_handler(self, name, flags):
         if name == "cflags" and self.spec.satisfies("+pic build_system=autotools"):
             flags.append(self.compiler.cc_pic_flag)
+        if name == "ldflags" and self.spec.satisfies("%cce@17"):
+            flags.append("-Wl,--undefined-version")
```
Why?!

Why cce@17 specifically? And why pass a linker arg depending on a compiler... it can be any linker?

Why would there be any unresolved versions?

It's even tested https://github.com/zlib-ng/zlib-ng/blob/2e3e5f30a05ee7b113c755ad9a075304e02bc328/.github/workflows/cmake.yml#L379

v-dobrev replied:
See the explanation here: https://github.com/spack/spack/pull/44010#discussion_r1628365884. Other compilers using LLD can be affected as well but I'm not sure, so I kept the compiler check to what I've tested.
